### ISSUE TITLE
.vo files vary from build order

### ISSUE THREAD

**Issue by bmwiedemann** [coq/coq#11229]
#### Description of the problem

While working on reproducible builds for openSUSE, I found that
In addition to #11227 , there are variations in .vo files
that go away when I build on a filesystem that has deterministic `readdir` order **and** build with `make` instead of `make -j4` - so that the builds happen in deterministic order.

```diff
/usr/lib64/coq/theories/FSets/FSetList.vo differs at offset '9059' (data)
--- /tmp/tmp.KtRUkpybQT/old.yPM 2019-12-02 10:39:36.148042421 +0000
+++ /tmp/tmp.KtRUkpybQT/new.vUR 2019-12-02 10:39:36.148042421 +0000
@@ -1,11 +1,11 @@
 00002340  8b e6 08 d0 e1 81 57 ec  a0 a0 28 4d 53 65 74 4c  |......W...(MSetL|
 00002350  69 73 74 a0 25 4d 53 65  74 73 a0 23 43 6f 71 40  |ist.%MSets.#Coq@|
-00002360  90 30 35 20 c8 a4 e3 1b  b6 aa ca 98 d2 da c7 7f  |.05 ............|
-00002370  f0 c1 a0 a0 29 4f 72 64  65 72 73 41 6c 74 a0 2a  |....)OrdersAlt.*|
+00002360  90 30 b5 b2 78 25 9a ee  e4 62 59 2b 46 34 9c 6c  |.0..x%...bY+F4.l|
+00002370  48 e4 a0 a0 29 4f 72 64  65 72 73 41 6c 74 a0 2a  |H...)OrdersAlt.*|
 00002380  53 74 72 75 63 74 75 72  65 73 a0 23 43 6f 71 40  |Structures.#Coq@|
 00002390  90 30 5c 91 d4 af b1 94  2f 7a 6f 13 17 a2 75 69  |.0\...../zo...ui|
-000023a0  ed 79 b1 cf 42 6c ca 2c  09 6f 38 78 f9 cc b9 cc  |.y..Bl.,.o8x....|
-000023b0  d4 7a 00 02 af 1c 84 95  a6 be 00 02 8b 52 00 00  |.z...........R..|
+000023a0  ed 79 6e 87 f6 25 58 27  a4 20 92 36 b7 1c 3b 0e  |.yn..%X'. .6..;.|
+000023b0  f3 06 00 02 af 1c 84 95  a6 be 00 02 8b 52 00 00  |.............R..|
```

https://github.com/bmwiedemann/openSUSE/blob/master/packages/c/coq/coq.spec#L77 has the details of how we build.

#### Coq Version

8.9.1

**Comment by ejgallego** [comment]
Umm, that's bizarre and certainly a bug; while we try to debug, do you have access to the differing files?

You can use `votour` [a binary in checker/] which is a vo file explorer to see what's exactly the different bit.

**Comment by SkySkimmer** [comment]
Since it looks like a Require issue it may be worth trying with 8.11/master.

**Comment by ejgallego** [comment]
`master` sees the same problem, I was able to reproduce.

**Comment by ejgallego** [comment]
Looks a bit tricky to analyze, I guess having the ability for `votour` to produce a diff would help.

I need to take care of other stuff, I reproduced just doing two dune builds, one with `-j 8` and the other with `-j 3`.

**Comment by ppedrot** [comment]
It might be due to the fact that hashconsing is not fully deterministic. Depending on the way the list of identifiers are processed we might get different in-memory representation.

**Comment by ejgallego** [comment]
> It might be due to the fact that hashconsing is not fully deterministic. Depending on the way the list of identifiers are processed we might get different in-memory representation.

Indeed I was wondering if the difference would be in the hash-cons, but how can this happen? Not for terms at least; almost surely what's causing the difference here is that we are taking some has of the directory contents.

**Comment by SkySkimmer** [comment]
In add_vo_path the directory names are put through `Id.of_string` which hconses in some arbitrary order. Maybe we should order the result of all_subdirs somehow.
(around https://github.com/coq/coq/blob/de91f71b2e25e66ba4dd1f1db6582f5fea205591/vernac/loadpath.ml#L246-L266)

**Comment by ejgallego** [comment]
> Maybe we should order the result of all_subdirs somehow.

I don't understand how this affects the hashconsing, tho even ordering would not yield the same results as for example some directories may not be present between two runs I think.

**Comment by bmwiedemann** [comment]
I could not find `votour` in openSUSE, so I uploaded 2 versions of one affected .vo file to https://rb.zq1.de/other/coq/ so that you can analyze it more closely.

**Comment by ppedrot** [comment]
@bmwiedemann the Coq version you used to generate these two files is still 8.9.1, right?

There is more than a change of representation, it seems that the provided files are different just because the vo files they are depending on are *already* different. Indeed, vo files store a digest of their dependencies and it happens in your example that it's essentially what is changing.

**Comment by ppedrot** [comment]
Forgot to mention, but the single offending file in your example is `Coq.MSets.MSetList`. For the file 1, it has digest `b5b278259aeee462592b46349c6c48e4` and for the other `3520c8a4e31bb6aaca98d2dac77ff0c1`.

**Comment by ejgallego** [comment]
@ppedrot indeed a few files are like that, but there is some root cause; I can provide a full `master` dump with different `.vo` files, but it is fairly easy to get it yourself:
```
$ git clean -fxd
$ make -f Makefile.dune DUNEOPT='-j8' # if you have an 8 core machine
$ cp -a _build/default/theories _theories1
$ find _build/default/theories/ -name '*.vo' | xargs rm
$ make -f Makefile.dune DUNEOPT='-j2'
$ cp -a _build/default/theories _theories2
$ find theories1 -name '*.vo' | xargs md5sum | sort -k 2 > v1
$ find theories2 -name '*.vo' | xargs md5sum | sort -k 2 > v2
$ diff -u v1 v2
```

**Comment by ppedrot** [comment]
I can provide a script to dump a structured representation of the binary contents of the vo, if that helps.

**Comment by ejgallego** [comment]
> I can provide a script to dump a structured representation of the binary contents of the vo, if that helps.

If that is diff-friendly would be great, I was thinking of coding a `vo` diff tool directly in OCaml.

[offtopic note, maybe it's about time we move from .vo, we should have some chat about it.]

**Comment by ppedrot** [comment]
BTW, I am spotting highly suspicious code in the safe demarshaller that seems to have been introduced by the native integer and float additions. I think we can get interesting segfaults when checking files from another architecture...

**Comment by ppedrot** [comment]
Here is a diff-friendly vo-dumping program: https://github.com/ppedrot/vodump.

**Comment by ppedrot** [comment]
I found a minimal counter-example using the commands provided by @ejgallego. Here is the [memory dump](https://github.com/coq/coq/files/3939439/dump.zip), where the first file was compiled with 4 cores and the second file with 2 cores. The two files differ on their library segment because they do not have the same memory representation even though AFAICT they are structurally equal. In particular, the one difference I could see is the object `0xed33` which corresponds to a `KerPair` value in turn contained in a `Const` term node. In the first case, it shares the string `"gcd"` with another part of the file, while in the second case it gets a fresh instance of the string. All other differences seem to be cause by this offset of size one.

Now, understanding why this happened is another story.

**Comment by ppedrot** [comment]
Note that this file uses modules, and they famously rely on an imperative delayed substitution mechanism...

**Comment by ppedrot** [comment]
I do suspect that this is linked to hashconsing, in a way that is unrelated to the directory contents. Rather, this is due to the fact that the GC itself is a source of randomness. My theory is the following:

1. We hashcons the string `"foo"` and store it in the global hashcons table.
2. The pointer to `"foo"` goes out of scope because the object that contained it is not reachable anymore
3. The GC may collect `"foo"` and **remove it from the hashcons table, or not** (← this is source of non-determinism)
4. We hashcons the string `"foo"` again. Now, depending on whether the GC has run, we get a fresh instance of the string or a mere backpointer.

Definitely the memory representation will be different depending on whether 3. occurred.

**Comment by SkySkimmer** [comment]
How does that break sharing though?

**Comment by ppedrot** [comment]
The issue appears right in a place where we use mutable state to delay the application of substitutions, so I guess it is an interaction with this imperative use of data.

**Comment by ejgallego** [comment]
Sounds quite plausible @ppedrot .

**Comment by gares** [comment]
What about optionally marshaling with the option `No_sharing` (to get reproducible builds now) and discuss at the next WG how to get rid of "randomized" hashconsing in full?

**Comment by ejgallego** [comment]
> What about optionally marshaling with the option `No_sharing` (to get reproducible builds now) and discuss at the next WG how to get rid of "randomized" hashconsing in full?

That could work, but to make it the default we'd have to bench size and time impact.

**Comment by SkySkimmer** [comment]
Is it confirmed that No_sharing fixes the issue?

**Comment by gares** [comment]
> Is it confirmed that No_sharing fixes the issue?

I did not try, but if @ppedrot is correct in his analysis then I believe it fixes the issue. Actually it could be a way to corroborate his thesis.

**Comment by ppedrot** [comment]
@gares I think we can expect trouble if we try this. This will be not far from not hashconsing at all, thus leading to memory explosion. This is still worth a shot though.

**Comment by bmwiedemann** [comment]
if you have a patch, I can test it with my https://github.com/bmwiedemann/reproducibleopensuse tools.

**Comment by ppedrot** [comment]
You can try [this branch](https://github.com/ppedrot/coq/tree/unshare-vo).

**Comment by bmwiedemann** [comment]
I just used this patch https://github.com/ppedrot/coq/commit/9be8ae861dac397da4076d3943ecbe5e43c35bfd and in the first build in a 8GB VM it ran out of memory in `theories/FSets/FMapFullAVL.vo`. With 30 GB it gave nice reproducible results.

**Comment by gares** [comment]
Nice!

Then the fix for the upcoming release should probably be to read `No_sharing` from an env variable.

This way distributions that need reproducibility of Coq (and Coq libraries) can just export such variable.
Final users may not care about reproducibility and get better performances on their files by not setting this variable (that would be the default).

**Comment by silene** [comment]
> Then the fix for the upcoming release should probably be to read `No_sharing` from an env variable.

How can that even be an acceptable fix? Coq and its standard library compile fine with only 2GB of RAM. If the minimal requirement grows to several tens of GB, this is quite a regression. And I am not even talking about the size of `.vo` files, which are presumably humongous, given the huge amount of RAM needed to (de)marshal them. The latter will impact all the users of distributions that decide to ship reproducible builds of Coq's standard library, even if those users do not especially care about it.

**Comment by gares** [comment]
I don't know how suse considers unreproducible packages. If it is just a warning, then they can live with it I guess.

I agree it is a quick fix and that the drawbacks are visible, and that a better solution should be looked for. But in the very short term, I don't have one.

**Comment by bmwiedemann** [comment]
It is a long-term goal to have reproducible builds. At some point I want to add a notification whenever an openSUSE package gets updated and is unreproducible. No automatic rejects soon.

So since you now have a better understanding of how the nondeterminism gets introduced, I suppose you can come up with a cleaner solution in some months? No need for dirty patches that have severe downsides.

**Comment by ppedrot** [comment]
Is there an easy way to use your infrastructure if we come up with some patches? Or should we just ask you?

**Comment by bmwiedemann** [comment]
You could use [my tools](https://github.com/bmwiedemann/reproducibleopensuse) on your hardware, or just use the simpler -j variations that ejgallego used.

I can also test patches for you (with the disadvantage of slower feedback cycles).

**Comment by ejgallego** [comment]
I think it is safe to remove this from 8.11 deadline, right @ppedrot ?

**Comment by bmwiedemann** [comment]
coq 8.11.2 is still affected.

**Comment by ejgallego** [comment]
> coq 8.11.2 is still affected.

Thanks for the report @bmwiedemann ; I'm afraid that 8.12 will still suffer from this; if @ppedrot analysis is correct the fix is not going to be easy due to the way Coq on-disk representation system is designed :S

**Comment by ppedrot** [comment]
Well, I do suspect that making module substitutions pure would solve the problem. This can be tried easily but since the bench infrastructure is currently down it's hard to check whether this would have negative performance consequences.

**Comment by ppedrot** [comment]
Seems to be fixed after #14337.

**Comment by silene** [comment]
> Seems to be fixed after #14337.

What is the idea? I have a hard time imagining delayed module substitution would be the only reason serialization would be non-deterministic. Was that the only source of unsharing in Coq?

**Comment by ppedrot** [comment]
@silene there is a tentative explanation a few comments above, although I don't know whether the problem is really solved or if I was not unlucky enough to hit it.

**Comment by bmwiedemann** [comment]
my test records show 8.11.1 or 8.12.0 as the first reproducible versions between 2020-04-10 and 2020-08-15

**Comment by silene** [comment]
> there is a tentative explanation a few comments above

I cannot find it. I understand why eagerly substituting modules is a step toward solving the issue, but I do not understand why it would definitely solve it. Is there no more hashconsing in Coq?

**Comment by ppedrot** [comment]
@silene the problem is not hashconsing, it hashconsing of mutable data structures. We don't have other instances of this phenomenon in the code. The only places I know where we have mutability is:
- the hashconsing of term arrays, but it doesn't matter precisely because this is synchronous with hashconsing
- delayed computation of hashes in names, but it probably doesn't matter because the mutable data is an integer

**Comment by Blaisorblade** [comment]
Since @bmwiedemann's records confirm Coq is now reproducible, tentatively closing.

**Comment by Blaisorblade** [comment]
Best reopened, based on https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Dune.20caching.20.26.20Coq.20non-reproducible.20builds.

**Comment by Blaisorblade** [comment]
Applying for Consortium support on behalf of Bedrock Systems (cc @gmalecha ).

**Comment by Blaisorblade** [comment]
@SkySkimmer Added as https://gist.githubusercontent.com/Blaisorblade/a7274f0eb15a13afbaa083b5b3673697/raw/f607062b1dab313da5118645f366bf0263409606/into_sep_at_term.v.

**Comment by SkySkimmer** [comment]
Going to move this to a new issue as I don't think the old comments are useful.

**Comment by ejgallego** [comment]
As this issue is still active, just a small note about `No_sharing`, using it in Coq is totally unfeasible, results in a crazy .vo size explosion (tried for some other experiments)
